File: python/best-time-to-buy-and-sell-stock.py
#  https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
#  Related Topics: Array, Dynamic Programming
#  Difficulty: Easy


# Initial thoughts:
# The brute force approach is to create a nested loop and check each pair of elements in the input array
# searching for the maximum profit

# Time complexity: O(n^2) where n === number of prices
# Space complexity: O(1)
from typing import List


class Solution:
    def maxProfitBruteForce(self, prices: List[int]) -> int:
        maxProfit = float('-inf')
        for i in range(len(prices)):
            for j in range(i+1, len(prices)):
                if prices[j] - prices[i] > maxProfit:
                    maxProfit = prices[j] - prices[i]
        return maxProfit if maxProfit > 0 else 0

# Optimization
# Since we are looking for the max of the differences between any prices[i+k] - prices[i]
# where 1 <= k < prices.length for each prices[i] it doesn't matter what the previous prices are.
# It is just enough to compare prices[i] with any prices[j] that comes after it.
# So moving forward we can keep the smallest prices[i] as minPrice and compare it with every
# following prices[j] and keep it if the difference is greater than previous differences.
# This will render our Time Complexity linear.

# Time complexity: O(n) where n === number of prices
# Space complexity: O(1)
    def maxProfit(self, prices: List[int]) -> int:
        if not len(prices):
            return 0
        maxProfit = float('-inf')
        minPrice = prices[0]
        for i in range(1, len(prices)):
            maxProfit = max(maxProfit, prices[i]-minPrice)
            minPrice = min(minPrice, prices[i])
        return maxProfit if maxProfit > 0 else 0

File: go/best_time_to_buy_and_sell_stock.go
package leetcode

// Time complexity: O(n) wheere n is the length of prices
// Space complexity: O(1)
func maxProfit(prices []int) int {
	min_price, max_profit := int(^uint(0)>>1), 0

	for _, el := range prices {
		max_profit = max(max_profit, el-min_price)
		min_price = min(min_price, el)
	}
	return max_profit

}

func max(x, y int) int {
	if x > y {
		return x
	} else {
		return y
	}
}

func min(x, y int) int {
	if x < y {
		return x
	} else {
		return y
	}
}
